packages/contracts/src/contracts/current/forwarder/MixinBuyExactTokens.sol

```diff
@@ -0,0 +1,128 @@
+        uint8 proxyId = uint8(orders[0].makerAssetData[0]);
+        require(proxyId == 1 || proxyId == 2, "unsupported token proxy");
+
+        uint256 remainingTakerTokenAmount = msg.value;
+        ETHER_TOKEN.deposit.value(remainingTakerTokenAmount)();
+        if (proxyId == 1) {
+            totalFillResults = buyExactTokensInternal(orders, signatures, feeOrders, feeSignatures, tokenAmount);
+        } else if (proxyId == 2) {
+            totalFillResults = buyNFTTokensInternal(orders, signatures, feeOrders, feeSignatures, remainingTakerTokenAmount);
+        }
+        remainingTakerTokenAmount = safeSub(remainingTakerTokenAmount, totalFillResults.takerAssetFilledAmount);
+        withdrawPayAndDeductFee(remainingTakerTokenAmount, totalFillResults.takerAssetFilledAmount, feeProportion, feeRecipient);
+        return totalFillResults;
+    }
+
+    function buyExactTokensInternal(
+        Order[] memory orders,
+        bytes[] memory signatures,
+        Order[] memory feeOrders,
+        bytes[] memory feeSignatures,
+        uint256 tokenAmount)
+        private
+        returns (Exchange.FillResults memory totalFillResults)
+    {
+        address makerTokenAddress = readAddress(orders[0].makerAssetData, 1); 
+        // We can short cut here for effeciency and use buyFeeTokensInternal if maker asset token is ZRX
+        // this buys us exactly that amount taking into account the fees saving gas
+        Exchange.FillResults memory requestedTokensResult;
+        if (makerTokenAddress == address(ZRX_TOKEN)) {
+            requestedTokensResult = buyFeeTokensInternal(orders, signatures, tokenAmount);
+            require(requestedTokensResult.makerAssetFilledAmount >= tokenAmount, "traded amount did not meet acceptable threshold");
+        } else {
+            Exchange.FillResults memory expectedMarketBuyFillResults = expectedMaketBuyFillResults(orders, tokenAmount);
+            if (expectedMarketBuyFillResults.takerFeePaid > 0) {
+                // Fees are required for these orders. Buy enough ZRX to cover the future market buy
+                Exchange.FillResults memory feeTokensResult = buyFeeTokensInternal(feeOrders, feeSignatures, expectedMarketBuyFillResults.takerFeePaid);
+                totalFillResults.takerAssetFilledAmount = feeTokensResult.takerAssetFilledAmount;
+                totalFillResults.takerFeePaid = feeTokensResult.takerFeePaid;
+            }
+            // Make our market buy of the requested tokens with the remaining balance
+            requestedTokensResult = EXCHANGE.marketBuyOrders(orders, tokenAmount, signatures);
+            require(requestedTokensResult.makerAssetFilledAmount == tokenAmount, "traded amount did not meet acceptable threshold");
+        }
+        addFillResults(totalFillResults, requestedTokensResult);
+        // Transfer all tokens to msg.sender
+        transferToken(makerTokenAddress, msg.sender, totalFillResults.makerAssetFilledAmount);
```

Comment:
Does sender abstraction allow us to perform these transfers in the Exchange? @abandeali1